src/store/index.js

```diff
@@ -34,25 +38,76 @@ export default new Vuex.Store({
       state.collections = await state.connection.db.listCollections().toArray()
     },
     async find({ state }, query = {}) {
-      state.totalNumberOfResults = null
+      if (state.collectionRecords[state.collectionName]) {
+        const query_serialization = JSON.stringify(query)
+
+        if (state.collectionRecords[state.collectionName].query === query_serialization) {
+          state.activeRecords = state.collectionRecords[state.collectionName]
+          state.totalNumberOfResults = state.activeRecords.result.length
+          return
+        }
+      }
+
+      state.totalNumberOfResult = null
       const collection = state.connection.db.collection(state.collectionName)
       const cursor = await collection.find(query)
+
       cursor.count((err, result) => {
         if (err) return
-
         state.totalNumberOfResults = result
       })
-      state.records = await cursor.toArray()
+
+      const result = await cursor.toArray()
+
+      state.activeRecords = {
+        query: JSON.stringify(query),
+        result,
+        isEmpty: !result.length
+      }
+
+      state.collectionRecords[state.collectionName] = state.activeRecords
     },
     async setDatabase({ state, dispatch }, databaseName) {
       state.connection = state.connection.useDb(databaseName)
       state.databaseName = databaseName
       state.collectionName = ''
-      state.records = []
+      state.collectionRecords = {}
+      state.openedCollections = []
       await dispatch('getCollections')
     },
     async setCollection({ state }, collectionName) {
       state.collectionName = collectionName
+      state.lastOpenedCollection = collectionName
+
+      if (state.collectionRecords[state.collectionName]) {
+        state.activeRecords = state.collectionRecords[state.collectionName]
+        state.totalNumberOfResults = state.activeRecords.result.length
+      } else {
+        state.activeRecords = []
+        state.totalNumberOfResults = null
+      }
+    },
+    async addCollectionToOpenedCollections({ state, dispatch }, collectionName) {
+      if (state.openedCollections.includes(collectionName)) {
+        return false
+      }
+
+      state.openedCollections.push(collectionName)
+      await dispatch('setCollection', collectionName)
+    },
+    async removeCollectionFromOpenedCollections({ state, dispatch }, collectionName) {
+      if (state.openedCollections.includes(collectionName)) {
+        state.openedCollections = state.openedCollections.filter(itemName => itemName !== collectionName)
+
+        //clear saved query records
+        delete state.collectionRecords[collectionName]
+
+        if (state.lastOpenedCollection === collectionName) {
+          await dispatch('setCollection', state.openedCollections[state.openedCollections.length - 1])
+        } else {
+          await dispatch('setCollection', state.lastOpenedCollection)
```

Comment:
prefer a single `await dispatch('setCollection'` here.

it could be something like: `await dispatch('setCollection', state.lastOpenedCollection == collectionName ? state.openedCollections[state.openedCollections.length - 1] : collectionName)`